src/lsp4clj/server.clj

```diff
@@ -199,21 +258,36 @@
                        join]
   protocols.endpoint/IEndpoint
   (start [this context]
-    (let [;; a thread so language server can use >!! and so that receive-message
-          ;; can use (>!! output-ch) to respect back pressure from clients that
-          ;; are slow to read.
-          pipeline (async/thread
-                     (loop []
-                       (if-let [message (async/<!! input-ch)]
-                         (do
-                           (receive-message this context message)
-                           (recur))
-                         (async/close! output-ch))))]
+    ;; Start receiving messages.
+    (let [pipeline (dispatch-input this context input-ch)]
+      ;; Wait to stop receiving messages.
       (async/go
-        ;; Wait for pipeline to close. This indicates input-ch was closed and
-        ;; that now output-ch is closed.
+        ;; When pipeline closes, it indicates input-ch has closed. We're done
+        ;; receiving.
         (async/<! pipeline)
-        ;; Do additional cleanup.
+        ;; Do cleanup.
```

Comment:
Any threads blocked indefinitely waiting for a response will never get unblocked again.

I think we should handle any `pending-sent-requests*` in this case, unblocking waiters with an error. We could implement that by putting a "poison pill` on the internal input channels, as right here and now we do not know if the response is still buffered.

Reply:
@ferdinand-beyer, so that I'm clear, this is a problem for Scarlet, but not for clojure-lsp, right? The difference being that clojure-lsp always shuts down the whole process after this cleanup is done, which ends up killing threads blocked waiting for a response. In contrast, Scarlet has a long-running server that manages the lifecycle of many lsp4clj servers. It shuts them down and restarts others without itself stopping entirely. So, it has to be more careful about thread management. Sound accurate?

I think you're worried, in particular, about threads waiting to `deref-or-cancel` or `deref` a pending-sent-request... is that correct? If so, it may be possible to `future-cancel` all those pending-sent-requests. That would `(deliver p ::cancelled)` to their promises, thus unblocking their waiters. If we do that, we might drop actual buffered client responses, a problem which the "poison pill" might avoid. But on the other hand, cancelling seems more direct, and since we're shutting down, the last few responses are probably irrelevant. (Basically, I'm not sure I could figure out how to implement a poison pill, but cancelling would be straightforward. 😃 )

Reply:
(And, FWIW, the lsp4clj docs are pretty clear that you should prefer `deref-or-cancel`, which means that blocked threads wouldn't stay blocked forever.)

Reply:
> this is a problem for Scarlet, but not for clojure-lsp, right?

Right now, yes. But I don't know who else is using lsp4clj, so there might be others in a similar situation. `deref-or-cancel` should indeed solve this, running into a timeout eventually.

I think `future-cancel` would cause more harm than good, as clients certainly don't expect to get `::cancelled` as a "magic" response value. If they use the 3-arity `deref` or `deref-or-cancel`, they would get their expected timeout value, not `::cancelled`.

The "poison pill" approach could indeed get ugly, as messages remaining in the buffer could cause a handler to send another request and waiting forever for the response, with the poison stuck in the buffer. It's definitely easier to just fail pending requests right away, using the `fail-pending-sent-requests` function I suggeted in the other thread.